### PR TITLE
Add new workflow to automatically (and periodically) update `ex_doc` version

### LINKED ISSUE #57: Trigger issue when new ex_doc releases have been released.
This should be doable with github actions I believe. It would make keeping up to date easier.

### PR DESCRIPTION
**Pull request by MarkoMin**
Tried everything locally (except actually creating an issue). If it works correctly, it will create one at midnight. First time having fun with GH actions so there may be some mistakes.

Closes #57

### CODE REVIEW

File: .github/workflows/ex_doc_version_check.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/ex_doc_version_check.yml:20]: This is probably safe, but if `ex_doc` ever releases a patch to an older version, there's no guarantee the latest tag is the "highest" in terms of versioning...

File: .github/workflows/ex_doc_version_check.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/ex_doc_version_check.yml:22]: This is also probably safe, if GitHub issues for `rebar3_ex_doc` don't "paginate" over a given limit, which is why I proposed the `EX_DOC_VSN` file.

File: .github/workflows/ex_doc_version_check.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/ex_doc_version_check.yml:29]: I wonder if this won't run in all forks, too (?)

Is there a way to further restrict this workflow? Otherwise forks either have to disable it by hand or have their "Actions" tab with issues every now and then (I'm assuming they can't just post an issue as simply as this, but have to use some kind of token).

File: .github/workflows/ex_doc_version_check.yml
Comment (MarkoMin) [reply .github/workflows/ex_doc_version_check.yml:20]: Well, yes, but as long as vMajor.minor.patch naming convention is used, it should be safe. We can utilize grep to only match versions which follow the convention

File: .github/workflows/ex_doc_version_check.yml
Comment (MarkoMin) [reply .github/workflows/ex_doc_version_check.yml:22]: Absolutely. We can solve this by fixing GH API version via header. I can submit modifications next week

File: .github/workflows/ex_doc_version_check.yml
Comment (MarkoMin) [reply .github/workflows/ex_doc_version_check.yml:29]: Yes, I believe it would. Prolly some env variable with current repo name exist, so we can wrap the whole task into IF clause.

File: .github/workflows/ex_doc_version_check.yml
Comment (paulo-ferraz-oliveira) [reply .github/workflows/ex_doc_version_check.yml:22]: > fixing GH API version via header

Does that prevent pagination? What I meant what that if you have too many issues, it's possible that the API won't give you all of them on the response and them you'll assume there's no issue (as per what you're searching) when there actually is one.

File: .github/workflows/ex_doc_version_check.yml
Comment (paulo-ferraz-oliveira) [reply .github/workflows/ex_doc_version_check.yml:29]: Or even better, not run the workflow in forks at all, otherwise it still pollutes the forks' tabs (though it might not be important 🤷).

Comment (paulo-ferraz-oliveira) [reply .github/workflows/ex_doc_version_check.yml:29]: It seems this changed recently, and now actions are opt-in, so I'll go ahead and resolve this one 👍 

image

File: .github/workflows/ex_doc_version_check.yml
Comment (MarkoMin) [reply .github/workflows/ex_doc_version_check.yml:22]: I think yes. Currently docs dont mention any pagination, but we can try and fetch issues from some repo that have few k issues just to be sure. So current version of API should be supported and work consistently for a few years

Comment (paulo-ferraz-oliveira) [reply .github/workflows/ex_doc_version_check.yml:22]: Yeah, doing this request on `erlang/otp`, as an example, brings up 30 results, while (atm) 275 are present...

Comment (paulo-ferraz-oliveira) [reply .github/workflows/ex_doc_version_check.yml:22]: One idea might be to tag the issue and then retrieve it based on that? e.g. tag it with `_ ex_doc_version_check` and then retrieve it with `labels= ex_doc_version_check` (?)

File: .github/workflows/ex_doc_version_check.yml
Comment (paulo-ferraz-oliveira) [reply .github/workflows/ex_doc_version_check.yml:20]: (mind you this API call's response is also paginated, getting only the first 30 results, as-is)

**Comment by paulo-ferraz-oliveira** [comment]
@MarkoMin, I created an Elixir script your pull request could plug in to (if it gets merged first), so that a pull request also gets created alongside the issue. 👍 

My pull request is mentioned above.

And here's an example of an automated pull request: https://github.com/paulo-ferraz-oliveira/rebar3_checkshell/blob/main/.github/workflows/rebar3_depup.sh (I do prefer to have my shell script separate so I can "shellcheck" and "shfmt" it 😄).

File: .github/workflows/ex_doc_version_check.yml
Comment (MarkoMin) [reply .github/workflows/ex_doc_version_check.yml:20]: Hmm, do we need more than the first page? But yeah, I completely missed pagination in this PR

Comment (paulo-ferraz-oliveira) [reply .github/workflows/ex_doc_version_check.yml:20]: > more than the first page?

Theoretically, yes; in practice I don't know. I just wanted to point out a potential issue.

Comment (MarkoMin) [reply .github/workflows/ex_doc_version_check.yml:20]: I can’t come up with an example where we would need them. I think we’re safe as long as the first page contains newest tags and if ex_doc doesn’t publish 2 releases in a day (in which case we’d miss one)

Comment (paulo-ferraz-oliveira) [reply .github/workflows/ex_doc_version_check.yml:20]: The example would be `ex_doc` to publish 30+ support versions for older tags (when paginating GitHub doesn't care about semver or whatever is "highest", just "more recent"). In that case (unlikely, but still...) even if your comparison were using semver it would fail to recognise a new version. And since I can't be sure there won't be 30+ support versions... now that I think about it, is there no way to get "latest"? That'd solve the pagination issue, too.

Comment (MarkoMin) [reply .github/workflows/ex_doc_version_check.yml:20]: Then we can increase per-page entries to considerably large number (e.g. 100) and sort tags by their name.

Comment (MarkoMin) [reply .github/workflows/ex_doc_version_check.yml:20]: okay, per-page parameter can me from 1-100, but I think it should do the job even if is not perfect...  Worst it can do is a false-positive issue that can be closed

**Comment by MarkoMin** [comment]
Oh, I didn't even think about that, that could work nicely! I'll utilize it if/when it gets merged.

File: .github/workflows/ex_doc_version_check.yml
Comment (paulo-ferraz-oliveira) [reply .github/workflows/ex_doc_version_check.yml]: I'm Ok with that.

File: .github/workflows/ex_doc_version_check.yml
Comment (paulo-ferraz-oliveira) [reply .github/workflows/ex_doc_version_check.yml]: They'd have to abuse the label alongside the title, right? I'm Ok with your current solution, though.

**Comment by MarkoMin** [comment]
One thought:  do we actually need both an issue and a PR? We should switch to PR only IMO.

**Comment by paulo-ferraz-oliveira** [comment]
I thought about that, too. Maybe PR alone is enough.

**Comment by MarkoMin** [comment]
Seems like it should be. But now we wait for #67 to get merged

**Comment by paulo-ferraz-oliveira** [comment]
@MarkoMin, #67's merged. It ended up being a Mix task. I think you'd need to compile it and run it, unless @starbelly has a different suggestion.

File: .github/workflows/ex_doc_version_check.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/ex_doc_version_check.yml:6]:
```suggestion
  workflow_dispatch:
  
```

File: .github/workflows/ex_doc_version_check.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/ex_doc_version_check.yml:20]:
```suggestion
                  jq -r ".[0].name"
```

Comment (paulo-ferraz-oliveira) [reply .github/workflows/ex_doc_version_check.yml:20]: `jq` is already installed in the Ubuntu runners...

File: .github/workflows/ex_doc_version_check.yml
Comment (paulo-ferraz-oliveira) [reply .github/workflows/ex_doc_version_check.yml:6]: This is to let maintainers run the workflow, from e.g. `main` by pressing a button and choosing a branch.

File: .github/workflows/ex_doc_version_check.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/ex_doc_version_check.yml:13]:
```suggestion
      - uses: actions/checkout@v4
```

File: .github/workflows/ex_doc_version_check.yml
Comment (paulo-ferraz-oliveira) [reply .github/workflows/ex_doc_version_check.yml:20]: As-was, though, this was returning `v0.9.0`, most probably due to the `sort`. Maybe since the list returns the latest tag as the first tag we should just get `per_page=1` (?)

File: .github/workflows/ex_doc_version_check.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/ex_doc_version_check.yml:1]: For consistency, we should probably name this workflow `up_ex_doc_version.yml`.

**Comment by starbelly** [comment]
A thought on this, and this may or may not be pragmatic, but we have both erlang and elixir at our disposal. We could write a bit of code that uses httpc to get this information and handle the result with more exactness (i.e., use verl or Version, Version if it's a mix task). 

Other thought, since we're wanting to get the latest release that is available on hex, we can just hit the hex api to get this info. 

```
$ latest_stable=$(curl -s https://hex.pm/api/packages/ex_doc | jq '.latest_stable_version' | tr -d '"')
$ echo $latest_stable
0.30.6
```

**Comment by paulo-ferraz-oliveira** [comment]
Good stuff. And `jq` has builtin support to remove the `"` too, so you can do `jq -r '.latest_stable_version'`. I'll suggest this above, instead of what I did before. I didn't quite get the Erlang idea, but feel free to expand. (e.g. you'd want the mix task to have no inputs and figure out the latest by itself?)

File: .github/workflows/ex_doc_version_check.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/ex_doc_version_check.yml:16]:
```suggestion
            CURR=`curl -s https://hex.pm/api/packages/ex_doc | jq -r '.latest_stable_version' \
```

Comment (paulo-ferraz-oliveira) [reply .github/workflows/ex_doc_version_check.yml:16]: As per @starbelly's comment on the main conversation.

**Comment by starbelly** [comment]
> Good stuff. And `jq` has builtin support to remove the `"` too, so you can do `jq -r '.latest_stable_version'`. I'll suggest this above, instead of what I did before. I didn't quite get the Erlang idea, but feel free to expand. (e.g. you'd want the mix task to have no inputs and figure out the latest by itself?)

Yeah. More like you can use erlang or elixir and httpc to do the gets and posts vs a bunch of shell scripterty, but that said. Let’s be pragmatic.

**Comment by MarkoMin** [comment]
I've made following changes:

1. Moved task from root to `lib/mix/tasks`
2. Renamed workflow as per @paulo-ferraz-oliveira suggestion
3. Update bash script to make an automated PR as per @paulo-ferraz-oliveira example - take a look if there are any commands that are redundant or missing

**Comment by paulo-ferraz-oliveira** [comment]
I approved the workflows to run. I'm forking your branch to test out the flow...

**Comment by paulo-ferraz-oliveira** [comment]
@MarkoMin, you copied to "lib/mix/tasks", instead of moving...

**Comment by paulo-ferraz-oliveira** [comment]
Example of a pull request working: https://github.com/paulo-ferraz-oliveira/rebar3_ex_doc_mm/pull/2/files.

File: .github/workflows/up_ex_doc_version.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/up_ex_doc_version.yml:15]:
```suggestion
      - uses: erlef/setup-beam@v1
        with:
          otp-version: 24
          elixir-version: 1.13
      - name: Get current version
```

Comment (paulo-ferraz-oliveira) [reply .github/workflows/up_ex_doc_version.yml:15]: Without this `mix` isn't available (it's used below).

Also, FYI, I picked the same Erlang/Elixir versions as from the other workflow.

File: .github/workflows/up_ex_doc_version.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/up_ex_doc_version.yml:18]:
```suggestion
            CURR=$(curl -s https://hex.pm/api/packages/ex_doc | jq -r '.latest_stable_version')
```

Comment (paulo-ferraz-oliveira) [reply .github/workflows/up_ex_doc_version.yml:18]: This command was incomplete (no closing "`"). Maybe it was a suggestion of mine, but here it's fixed.

File: .github/workflows/up_ex_doc_version.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/up_ex_doc_version.yml:19]:
```suggestion
            TITLE="[automation] Update \`ex_doc\` to ${CURR}"
```

File: .github/workflows/up_ex_doc_version.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/up_ex_doc_version.yml:20]:
```suggestion
            PR=$(curl -s https://api.github.com/repos/${GITHUB_REPOSITORY}/pulls?state=all%26per_page=100 | jq ".[] | select(.title==\"${TITLE}\")")
```

Comment (paulo-ferraz-oliveira) [reply .github/workflows/up_ex_doc_version.yml:20]: I ease running in forks (by means of `$GITHUB_REPOSITORY`) and also use `jq` instead of `grep`.

File: .github/workflows/up_ex_doc_version.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/up_ex_doc_version.yml:22]:
```suggestion
            echo "Searching for pull request with title ${TITLE}..."
            
            if [ -z "${PR}" ]
```

Comment (paulo-ferraz-oliveira) [reply .github/workflows/up_ex_doc_version.yml:22]: Logging and minor change as per what was suggested above...

File: .github/workflows/up_ex_doc_version.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/up_ex_doc_version.yml:26]:
```suggestion
              echo "  ... not found in 100 most recent pull requests!"
```

Comment (paulo-ferraz-oliveira) [reply .github/workflows/up_ex_doc_version.yml:26]: Logging...

File: .github/workflows/up_ex_doc_version.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/up_ex_doc_version.yml:28]:
```suggestion
              git fetch origin
              
```

Comment (paulo-ferraz-oliveira) [reply .github/workflows/up_ex_doc_version.yml:28]: If we don't do it here branch comparison is wrong.

File: .github/workflows/up_ex_doc_version.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/up_ex_doc_version.yml:27]:
```suggestion
              git fetch origin
              
              BRANCH="up_ex_doc_to_${CURR}"
```

Comment (paulo-ferraz-oliveira) [reply .github/workflows/up_ex_doc_version.yml:27]: If we don't do it (`fetch`), here, branch comparison is wrong.

File: .github/workflows/up_ex_doc_version.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/up_ex_doc_version.yml:30]:
```suggestion
                  echo "Branch $BRANCH found! Bailing out..."
```

Comment (paulo-ferraz-oliveira) [reply .github/workflows/up_ex_doc_version.yml:30]: Logging...

File: .github/workflows/up_ex_doc_version.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/up_ex_doc_version.yml:34]:
```suggestion
```

Comment (paulo-ferraz-oliveira) [reply .github/workflows/up_ex_doc_version.yml:34]: Moved above.

File: .github/workflows/up_ex_doc_version.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/up_ex_doc_version.yml:36]:
```suggestion
              mix deps.get
              mix compile
```

Comment (paulo-ferraz-oliveira) [reply .github/workflows/up_ex_doc_version.yml:36]: Without `mix deps.get` this fails...

File: .github/workflows/up_ex_doc_version.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/up_ex_doc_version.yml:40]:
```suggestion
                  echo "There's stuff to push. Creating a pull request..."
```

Comment (paulo-ferraz-oliveira) [reply .github/workflows/up_ex_doc_version.yml:40]: Logging...

File: .github/workflows/up_ex_doc_version.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/up_ex_doc_version.yml:42]:
```suggestion
                  git commit -m "${TITLE}"
```

Comment (paulo-ferraz-oliveira) [reply .github/workflows/up_ex_doc_version.yml:42]: Consistency...

File: .github/workflows/up_ex_doc_version.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/up_ex_doc_version.yml:46]:
```suggestion
                      --title "${TITLE}" \
```

Comment (paulo-ferraz-oliveira) [reply .github/workflows/up_ex_doc_version.yml:46]: Consistency...

File: .github/workflows/up_ex_doc_version.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/up_ex_doc_version.yml:47]:
```suggestion
                      --body "This is an automated action to update \`ex_doc\` to version ${CURR}"
```

File: .github/workflows/up_ex_doc_version.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/up_ex_doc_version.yml:48]:
```suggestion
              else
                  echo "Nothing to push. Is \`ex_doc\` already at version ${CURR}?"
              fi
            else
              echo "  ... found! Bailing out..."
```

File: .github/workflows/up_ex_doc_version.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/up_ex_doc_version.yml:49]:
```suggestion
            fi
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

Comment (paulo-ferraz-oliveira) [reply .github/workflows/up_ex_doc_version.yml:49]: Without this you can't push commits or create pull requests. More info. here: https://docs.github.com/en/actions/security-guides/automatic-token-authentication

File: .github/workflows/up_ex_doc_version.yml
Comment (paulo-ferraz-oliveira) [inline .github/workflows/up_ex_doc_version.yml:16]:
```suggestion
```

File: lib/mix/tasks/up_ex_doc_version.ex
Comment (paulo-ferraz-oliveira) [inline lib/mix/tasks/up_ex_doc_version.ex:15]:
```suggestion
    env = [{"PATH", System.get_env("PATH")}]
    System.cmd("mix", ["deps.unlock", "ex_doc"], env: env)
    System.cmd("mix", ["deps.get"], env: env)
```

Comment (paulo-ferraz-oliveira) [reply lib/mix/tasks/up_ex_doc_version.ex:15]: I had an off-by-one issue here. Mix's state is not updated when we write the file, so `Mix.Task.run` would only work on the second update.

File: lib/mix/tasks/up_ex_doc_version.ex
Comment (paulo-ferraz-oliveira) [inline lib/mix/tasks/up_ex_doc_version.ex:1]: - [ ] you need to remove `up_ex_doc_version.ex` at the root folder.

**Comment by paulo-ferraz-oliveira** [comment]
@MarkoMin, sorry I did so many small changes/suggestions, but it was the best way I found to explain my reasoning. Accepting them all should put the pull request in a mergeable state.

**Comment by starbelly** [comment]
> @starbelly, if this isn't done already we're gonna need to enable
> image
> 
> and also
> image

I think with how infrequent ex_doc releases are (and should be) we should just allow the bot to open PR. We definitely want to see what the bot has done before merging, thoughts?

**Comment by paulo-ferraz-oliveira** [comment]
That's my goal. I was under the impression you needed:

- the latter to allow to write to branch
- the former to open the pull request

But maybe the former is not required (?)

**Comment by paulo-ferraz-oliveira** [comment]
We can also accept+merge (once all changes are done), then test the settings that are required (our goal is only to create the pull request, as automating stuff like this can lead to issues - let's say that `ex_doc` breaks and we don't have enough tests, for example).

**Comment by MarkoMin** [comment]
I approved all suggestions. Sorry for errors with `mix`, first time using it. 

Now it should be done. It's quite messy with all those commits, but we can squash them I guess. For some reason, `mix.exs` got into commits, but changed version is the file that is currently on main so no conflicts should exists.

File: lib/mix/tasks/up_ex_doc_version.ex
Comment (starbelly) [inline lib/mix/tasks/up_ex_doc_version.ex:15]: We can do this and it should be fine, but probably more pragmatic is to run this, then execute `mix deps.get`  back in the shell (i.e., in whatever workflow that will use this). 

Thoughts?

Comment (paulo-ferraz-oliveira) [reply lib/mix/tasks/up_ex_doc_version.ex:15]: I don't mind doing it in another pull request @starbelly, even before we release. This fix is already scope creep caused by me, since it wasn't in the change proposed by @MarkoMin initially.

**Comment by paulo-ferraz-oliveira** [comment]
@MarkoMin, could you rebase this onto the main branch, and move the script?

I see this in the Files' tab

image

and it's confusing. Thanks.

**Comment by MarkoMin** [comment]
> @MarkoMin, could you rebase this onto the main branch, and move the script?

done

File: lib/mix/tasks/up_ex_doc_version.ex
Comment (paulo-ferraz-oliveira) [reply lib/mix/tasks/up_ex_doc_version.ex:15]: @MarkoMin, if you prefer to do it, you can remove these calls to mix from here and set them in the CI bit... But lemme know, otherwise I can also do in a subsequent pr (since this one's working as-is)

Comment (MarkoMin) [reply lib/mix/tasks/up_ex_doc_version.ex:15]: I can do it. Do I need to remove File:write also or just lines 14 and 15?

Comment (MarkoMin) [reply lib/mix/tasks/up_ex_doc_version.ex:15]: done

**Review by starbelly — APPROVED** [review]
LGTM! If it fails, we can fix it 😄

**Pull request merged**
